src/sqlfluff/core/linter/linting_result.py

```diff
@@ -181,3 +184,34 @@ def tree(self) -> Optional[BaseSegment]:  # pragma: no cover
                 "path."
             )
         return self.paths[0].tree
+
+    def check_templating_or_parse_errors(self) -> Tuple[int, int]:
+        """Treat tiles with templating or parse error as unfixable.
+
+        NOTE! THIS FUNCTION HAS SIDE EFFECTS: IT DISCARDS LINT FIXES.
+
+        Scan all LintedFiles:
+        - Files with none of these errors: No action
+        - Files with these errors (before filtering): Removes fixes from
+          SQLLintError objects, i.e. marks them "unfixable"
+
+        Returns the total number of these errors before and after "ignore/noqa"
+        filters are applied.
+        """
+        types = (SQLParseError, SQLTemplaterError)
+        total_errors = self.num_violations(types=types, filter_ignore=False)
+        num_filtered_errors = 0
+        if total_errors:
+            for linted_dir in self.paths:
+                for linted_file in linted_dir.files:
+                    num_errors = linted_file.num_violations(
+                        types=types, filter_ignore=False
+                    )
+                    if num_errors:
+                        # File has errors. Discard all the fixes: they are
+                        # potentially unsafe.
+                        for violation in linted_file.violations:
+                            if hasattr(violation, "fixes"):
+                                violation.fixes = []  # type: ignore
+                        num_filtered_errors += linted_file.num_violations(types=types)
+        return total_errors, num_filtered_errors
```

Comment:
This function is the main part of the PR. Every flavor of "fix" eventually uses it:
* `fix` command line
* `fix` command line for stdin (special case)
* Simple API `fix()` function